# WeRateDogs Tweeter Archive

## Gather

In [2]:
import pandas as pd
import numpy as np
import requests
#import tweepy

In [3]:
tweet = pd.read_csv('twitter-archive-enhanced.csv')

In [4]:
###
r = requests.get('https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv')
if r.status_code == 200:
    with open('image_predictions.tsv', 'wb') as f:
        f.write(r.content)

In [5]:
image = pd.read_csv('image_predictions.tsv', sep = '\t')

In [7]:
###
consumer_key = 'X4Mu5bFyMBpjjstVS4XbYjZcL'
consumer_secret = '6nXJBPj6kBPzT0zP45fZnjiAOuMTN5OEBS3tQ7xKbfIT4NuIst'
access_token = '4366384514-l6MKA0qiPnv018rq4aSQPQd5TODG6Ko4gNeBlSA'
access_secret = '6HcJzhQoY9fZaTfj8p5XXyBdCjNsKiB8SUURkRmSQEh3H'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit = True, wait_on_rate_limit_notify = True )

In [12]:
###
from timeit import default_timer as timer

start = timer()
tweet_id = []
retweet_count = []
favorite_count = []

for ids in image.tweet_id:
    try:
        tweet = api.get_status(ids, tweet_mode='extended')
        tweet_id.append(ids)
        retweet_count.append(tweet.retweet_count)
        favorite_count.append(tweet.favorite_count)
    except Exception as e:
        print(str(e))
        
end = timer()
print(end - start)    

Rate limit reached. Sleeping for: 256
[{u'message': u'No status found with that ID.', u'code': 144}]
Rate limit reached. Sleeping for: 732
[{u'message': u'No status found with that ID.', u'code': 144}]
[{u'message': u'No status found with that ID.', u'code': 144}]
[{u'message': u'No status found with that ID.', u'code': 144}]
[{u'message': u'No status found with that ID.', u'code': 144}]
[{u'message': u'No status found with that ID.', u'code': 144}]
1383.16354203


In [14]:
###
file = open("tweet_json.txt", "w")
for index in range(len(tweet_id)):
    file.write(str(tweet_id[index]) + " " + str(retweet_count[index]) + " " + str(favorite_count[index]) + "\n")
file.close()

In [6]:
info = pd.read_csv('tweet_json.txt', sep=" ", header=None)
info.columns = ['tweet_id','retweet_count','favorite_count']

## Assess

In [7]:
tweet.head(10)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  8.924210e+17                    NaN                  NaN   
1  8.921770e+17                    NaN                  NaN   
2  8.918150e+17                    NaN                  NaN   
3  8.916900e+17                    NaN                  NaN   
4  8.913280e+17                    NaN                  NaN   
5  8.910880e+17                    NaN                  NaN   
6  8.909720e+17                    NaN                  NaN   
7  8.907290e+17                    NaN                  NaN   
8  8.906090e+17                    NaN                  NaN   
9  8.902400e+17                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   
5  2017-07-29 00:08:17 +0000   
6  2017-07-28 16:27:12 +0000   
7  2017-07-28 00:22:40 +0000   
8  2017-07-27 16:25:51 +0000   
9  2017-07-26 15:59:51 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   
5  <a href="http://twitter.com/download/iphone" r...   
6  <a href="http://twitter.com/download/iphone" r...   
7  <a href="http://twitter.com/download/iphone" r...   
8  <a href="http://twitter.com/download/iphone" r...   
9  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   
5  Here we have a majestic great white breaching ...                  NaN   
6  Meet Jax. He enjoys ice cream so much he gets ...                  NaN   
7  When you watch your owner call another dog a g...                  NaN   
8  This is Zoey. She doesn't want to be one of th...                  NaN   
9  This is Cassie. She is a college pup. Studying...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   
5                       NaN                        NaN   
6                       NaN                        NaN   
7                       NaN                        NaN   
8                       NaN                        NaN   
9                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   
5  https://twitter.com/dog_rates/status/891087950...                13   
6  https://gofundme.com/ydvmve-surgery-for-jax,ht...                13   
7  https://twitter.com/dog_rates/status/890729181...                13   
8  https://twitter.com/dog_rates/status/890609185...                13   
9  https://twitter.com/dog_rates/status/890240255...                14   

   rating_denominator      name  doggo fl

In [8]:
image.head(10)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5  666050758794694657  https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6  666051853826850816  https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7  666055525042405380  https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8  666057090499244032  https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9  666058600524156928  https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   
5        1    Bernese_mountain_dog  0.651137    True    English_springer   
6        1              box_turtle  0.933012   False          mud_turtle   
7        1                    chow  0.692517    True     Tibetan_mastiff   
8        1           shopping_cart  0.962465   False     shopping_basket   
9        1        miniature_poodle  0.201493    True            komondor   

    p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0  0.156665    True            Shetland_sheepdog  0.061428    True  
1  0.074192    True          Rhodesian_ridgeback  0.072010    True  
2  0.138584    True                   bloodhound  0.116197    True  
3  0.360687    True           miniature_pinscher  0.222752    True  
4  0.243682    True                     Doberman  0.154629    True  
5  0.263788    True   Greater_Swiss_Mountain_dog  0.016199    True  
6  0.045885   False                     terrapin  0.017885   False  
7  0.058279    True                     fur_coat  0.054449   False  
8  0.014594   False             golden_retriever  0.007959    True  
9  0.192305    True  soft-coated_wheaten_terrier  0.082086    True

In [9]:
info.head(10)

tweet_id  retweet_count  favorite_count
0  666020888022790149            519            2521
1  666029285002620928             47             131
2  666033412701032449             45             125
3  666044226329800704            141             302
4  666049248165822465             40             109
5  666050758794694657             59             133
6  666051853826850816            864            1236
7  666055525042405380            258             441
8  666057090499244032            143             299
9  666058600524156928             58             114

In [10]:
tweet.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null float64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(5), int64(2), 

In [11]:
image.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.0+ KB


In [12]:
info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2069 entries, 0 to 2068
Data columns (total 3 columns):
tweet_id          2069 non-null int64
retweet_count     2069 non-null int64
favorite_count    2069 non-null int64
dtypes: int64(3)
memory usage: 48.6 KB


In [13]:
all_columns = pd.Series(list(tweet) + list(image) + list(info))
all_columns[all_columns.duplicated()]

17    tweet_id
29    tweet_id
dtype: object

In [14]:
list(tweet)

['tweet_id',
 'in_reply_to_status_id',
 'in_reply_to_user_id',
 'timestamp',
 'source',
 'text',
 'retweeted_status_id',
 'retweeted_status_user_id',
 'retweeted_status_timestamp',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'doggo',
 'floofer',
 'pupper',
 'puppo']

In [15]:
list(image)

['tweet_id',
 'jpg_url',
 'img_num',
 'p1',
 'p1_conf',
 'p1_dog',
 'p2',
 'p2_conf',
 'p2_dog',
 'p3',
 'p3_conf',
 'p3_dog']

In [16]:
tweet[tweet['expanded_urls'].isnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
30    8.862670e+17           8.860000e+17         2.281182e+09   
55    8.816330e+17           8.820000e+17         4.738443e+07   
64    8.796740e+17           8.800000e+17         3.105441e+09   
113   8.707260e+17           8.710000e+17         1.648776e+07   
148   8.634280e+17           8.630000e+17         7.759620e+07   
179   8.572150e+17           8.570000e+17         1.806710e+08   
185   8.563310e+17                    NaN                  NaN   
186   8.562880e+17           8.560000e+17         2.792810e+08   
188   8.558630e+17           8.560000e+17         1.943518e+08   
189   8.558600e+17           8.560000e+17         1.361572e+07   
218   8.503340e+17           8.500000e+17         2.195506e+07   
228   8.482140e+17           8.480000e+17         4.196984e+09   
234   8.476170e+17           8.480000e+17         4.196984e+09   
274   8.406990e+17           8.410000e+17         8.410000e+17   
290   8.381500e+17           8.380000e+17         2.195506e+07   
291   8.380860e+17           8.380000e+17         2.894131e+09   
313   8.352460e+17           8.350000e+17         2.625958e+07   
342   8.320890e+17           8.320000e+17         3.058208e+07   
346   8.319270e+17           8.320000e+17         2.068372e+07   
375   8.283620e+17                    NaN                  NaN   
387   8.265990e+17           8.270000e+17         4.196984e+09   
409   8.233330e+17           8.230000e+17         1.582854e+09   
427   8.211530e+17           8.210000e+17         1.132119e+08   
498   8.131300e+17           8.130000e+17         4.196984e+09   
513   8.116480e+17           8.120000e+17         4.196984e+09   
570   8.018550e+17           8.020000e+17         1.185634e+07   
576   8.008590e+17           8.010000e+17         2.918590e+08   
611   7.971660e+17           7.970000e+17         2.916630e+07   
701   7.860510e+17           7.730000e+17         7.310000e+17   
707   7.855150e+17                    NaN                  NaN   
843   7.667150e+17           7.670000e+17         4.196984e+09   
857   7.639570e+17           7.640000e+17         1.584641e+07   
967   7.503820e+17           7.500000e+17         4.717297e+09   
1005  7.476510e+17           7.480000e+17         4.196984e+09   
1080  7.388910e+17           7.380000e+17         3.589728e+08   
1295  7.079830e+17           7.080000e+17         2.319108e+09   
1345  7.044910e+17           7.040000e+17         2.878549e+07   
1445  6.965180e+17                    NaN                  NaN   
1446  6.964910e+17           6.960000e+17         4.196984e+09   
1474  6.936440e+17           6.940000e+17         4.196984e+09   
1479  6.935820e+17           6.940000e+17         1.198989e+09   
1497  6.924230e+17           6.920000e+17         4.196984e+09   
1523  6.906070e+17           6.900000e+17         4.670367e+08   
1598  6.860360e+17           6.860000e+17         4.196984e+09   
1605  6.856810e+17           6.860000e+17         4.196984e+09   
1618  6.849700e+17           6.850000e+17         4.196984e+09   
1663  6.828090e+17           6.830000e+17         4.196984e+09   
1689  6.813410e+17           6.810000e+17         4.196984e+09   
1774  6.780230e+17           6.780000e+17         4.196984e+09   
1819  6.765910e+17           6.770000e+17         4.196984e+09   
1844  6.758490e+17           6.760000e+17         4.196984e+09   
1895  6.747430e+17           6.750000e+17         4.196984e+09   
1905  6.746070e+17           6.740000e+17         4.196984e+09   
1914  6.743310e+17           6.660000e+17         1.637468e+07   
1940  6.737160e+17           6.740000e+17         4.196984e+09   
2038  6.715500e+17           6.720000e+17         4.196984e+09   
2149  6.696850e+17           6.690000e+17         4.196984e+09   
2189  6.689680e+17           6.690000e+17         2.143566e+07   
2298  6.670700e+17           6.670000e+17         4.196984e+09   

                      timestamp  \
30    2017-07-

In [18]:
tweet.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455128e+17         2.015385e+16   
std    6.856706e+16           7.583419e+16         1.253546e+17   
min    6.660210e+17           6.660000e+17         1.185634e+07   
25%    6.783992e+17           6.760000e+17         3.086374e+08   
50%    7.196275e+17           7.035000e+17         4.196984e+09   
75%    7.993375e+17           8.260000e+17         4.196984e+09   
max    8.924210e+17           8.860000e+17         8.410000e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720221e+17              1.241437e+16         13.126486   
std           6.236131e+16              9.597227e+16         45.876648   
min           6.660000e+17              7.832140e+05          0.000000   
25%           7.190000e+17              4.196984e+09         10.000000   
50%           7.800000e+17              4.196984e+09         11.000000   
75%           8.200000e+17              4.196984e+09         12.000000   
max           8.870000e+17              7.870000e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [19]:
image.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [20]:
info.describe()

tweet_id  retweet_count  favorite_count
count  2.069000e+03    2069.000000     2069.000000
mean   7.381251e+17    2895.317061     8575.327211
std    6.766215e+16    4891.074703    12506.787889
min    6.660209e+17      13.000000        0.000000
25%    6.764400e+17     618.000000     1648.000000
50%    7.117327e+17    1373.000000     3808.000000
75%    7.931355e+17    3365.000000    10822.000000
max    8.924206e+17   77903.000000   144091.000000

In [21]:
tweet.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1090  7.376790e+17                    NaN                  NaN   
1086  7.381660e+17                    NaN                  NaN   
2032  6.717630e+17                    NaN                  NaN   
15    8.892790e+17                    NaN                  NaN   
2037  6.715610e+17                    NaN                  NaN   

                      timestamp  \
1090  2016-05-31 16:14:39 +0000   
1086  2016-06-02 00:32:39 +0000   
2032  2015-12-01 18:50:38 +0000   
15    2017-07-24 00:19:32 +0000   
2037  2015-12-01 05:26:34 +0000   

                                                 source  \
1090  <a href="http://twitter.com/download/iphone" r...   
1086  <a href="http://twitter.com/download/iphone" r...   
2032  <a href="http://twitter.com/download/iphone" r...   
15    <a href="http://twitter.com/download/iphone" r...   
2037  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1090  We only rate dogs. Pls stop sending in non-can...                  NaN   
1086  This is Axel. He's a professional leaf catcher...                  NaN   
2032  Say hello to Mark. He's a good dog. Always rea...                  NaN   
15    This is Oliver. You're witnessing one of his m...                  NaN   
2037  This is the best thing I've ever seen so sprea...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1090                       NaN                        NaN   
1086                       NaN                        NaN   
2032                       NaN                        NaN   
15                         NaN                        NaN   
2037                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1090  https://twitter.com/dog_rates/status/737678689...                11   
1086  https://twitter.com/dog_rates/status/738166403...                12   
2032  https://twitter.com/dog_rates/status/671763349...                 9   
15    https://twitter.com/dog_rates/status/889278841...                13   
2037  https://twitter.com/dog_rates/status/671561002...                13   

      rating_denominator    name doggo floofer pupper puppo  
1090                  10    None  None    None   None  None  
1086                  10    Axel  None    None   None  None  
2032                  10    Mark  None    None   None  None  
15                    10  Oliver  None    None   None  None  
2037                  10     the  None    None   None  None

In [22]:
tweet.name.value_counts()

None         745
a             55
Charlie       12
Cooper        11
Lucy          11
Oliver        11
Tucker        10
Penny         10
Lola          10
Bo             9
Winston        9
Sadie          8
the            8
Daisy          7
Bailey         7
Toby           7
an             7
Buddy          7
Jack           6
Leo            6
Scout          6
Rusty          6
Milo           6
Oscar          6
Jax            6
Koda           6
Stanley        6
Bella          6
Dave           6
Finn           5
            ... 
Glacier        1
Pip            1
Zoe            1
Cilantro       1
Clifford       1
Huxley         1
Jennifur       1
Lenox          1
Kane           1
Jeremy         1
Zooey          1
Todo           1
Kanu           1
Ember          1
Steve          1
Laika          1
Socks          1
Eazy           1
Finnegus       1
Zeek           1
Alexander      1
Meatball       1
Zara           1
Bronte         1
Edgar          1
Glenn          1
Ester          1
Obi           

In [23]:
tweet.source.value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       33
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

In [31]:
image.jpg_url.value_counts()

https://pbs.twimg.com/media/CUN4Or5UAAAa5K4.jpg                                            2
https://pbs.twimg.com/media/CdHwZd0VIAA4792.jpg                                            2
https://pbs.twimg.com/media/CzG425nWgAAnP7P.jpg                                            2
https://pbs.twimg.com/media/CtKHLuCWYAA2TTs.jpg                                            2
https://pbs.twimg.com/media/CVuQ2LeUsAAIe3s.jpg                                            2
https://pbs.twimg.com/media/Ct72q9jWcAAhlnw.jpg                                            2
https://pbs.twimg.com/ext_tw_video_thumb/815965888126062592/pu/img/JleSw4wRhgKDWQj5.jpg    2
https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg                                            2
https://pbs.twimg.com/media/CWza7kpWcAAdYLc.jpg                                            2
https://pbs.twimg.com/media/CU1zsMSUAAAS0qW.jpg                                            2
https://pbs.twimg.com/media/CwiuEJmW8AAZnit.jpg                       

In [26]:
tweet[tweet.name.duplicated()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
7     8.907290e+17                    NaN                  NaN   
12    8.896650e+17                    NaN                  NaN   
23    8.874740e+17                    NaN                  NaN   
24    8.873430e+17                    NaN                  NaN   
25    8.871010e+17                    NaN                  NaN   
30    8.862670e+17           8.860000e+17         2.281182e+09   
32    8.860540e+17                    NaN                  NaN   
35    8.855190e+17                    NaN                  NaN   
37    8.851680e+17                    NaN                  NaN   
41    8.844420e+17                    NaN                  NaN   
42    8.842480e+17                    NaN                  NaN   
47    8.831180e+17                    NaN                  NaN   
55    8.816330e+17           8.820000e+17         4.738443e+07   
59    8.808720e+17                    NaN                  NaN   
60    8.804660e+17                    NaN                  NaN   
62    8.800960e+17                    NaN                  NaN   
64    8.796740e+17           8.800000e+17         3.105441e+09   
72    8.786050e+17                    NaN                  NaN   
75    8.782820e+17                    NaN                  NaN   
76    8.780580e+17                    NaN                  NaN   
78    8.776110e+17                    NaN                  NaN   
83    8.765380e+17                    NaN                  NaN   
88    8.750970e+17                    NaN                  NaN   
89    8.750210e+17                    NaN                  NaN   
93    8.740580e+17                    NaN                  NaN   
96    8.735800e+17                    NaN                  NaN   
98    8.732140e+17                    NaN                  NaN   
99    8.729670e+17                    NaN                  NaN   
100   8.728210e+17                    NaN                  NaN   
101   8.726690e+17                    NaN                  NaN   
...            ...                    ...                  ...   
2326  6.664120e+17                    NaN                  NaN   
2327  6.664070e+17                    NaN                  NaN   
2328  6.663960e+17                    NaN                  NaN   
2329  6.663740e+17                    NaN                  NaN   
2330  6.663630e+17                    NaN                  NaN   
2331  6.663530e+17                    NaN                  NaN   
2332  6.663450e+17                    NaN                  NaN   
2333  6.663380e+17                    NaN                  NaN   
2334  6.662940e+17                    NaN                  NaN   
2335  6.662870e+17                    NaN                  NaN   
2336  6.662730e+17                    NaN                  NaN   
2337  6.662690e+17                    NaN                  NaN   
2338  6.661040e+17                    NaN                  NaN   
2339  6.661020e+17                    NaN                  NaN   
2340  6.661000e+17                    NaN                  NaN   
2341  6.660940e+17                    NaN                  NaN   
2342  6.660830e+17                    NaN                  NaN   
2343  6.660730e+17                    NaN                  NaN   
2344  6.660710e+17                    NaN                  NaN   
2345  6.660640e+17                    NaN                  NaN   
2346  6.660590e+17                    NaN                  NaN   
2347  6.660570e+17                    NaN                  NaN   
2348  6.660560e+17                    NaN                  NaN   
2349  6.660520e+17                    NaN                  NaN   
2350  6.660510e+17                    NaN                  NaN   
2351  6.660490e+17                    NaN                  NaN   
2352  6.660440e+17                    NaN                  NaN   
2353  6.660330e+17                    NaN                  NaN   
2354  6.660290e+17                    NaN         

In [27]:
tweet.rating_numerator.sort_values()

315        0
1016       0
2335       1
2261       1
2338       1
605        1
1446       1
1869       1
2091       1
2038       1
1940       1
1764       2
2310       2
1761       2
1920       2
2246       2
2237       2
2349       2
2326       2
2079       2
2305       3
1249       3
1898       3
1478       3
1692       3
2202       3
765        3
1928       3
1314       3
2183       3
        ... 
866       14
83        14
291       15
285       15
55        17
1663      20
516       24
1712      26
763       27
1433      44
1274      45
1202      50
1351      60
340       75
695       75
1254      80
433       84
1843      88
1228      99
1635     121
1634     143
1779     144
902      165
290      182
1120     204
2074     420
188      420
189      666
313      960
979     1776
Name: rating_numerator, Length: 2356, dtype: int64

In [28]:
tweet.rating_denominator.sort_values()

313       0
2335      2
516       7
1576     10
1575     10
1574     10
1573     10
1572     10
1563     10
1570     10
1569     10
1568     10
1567     10
1566     10
1565     10
1564     10
1577     10
1571     10
0        10
1562     10
1547     10
1548     10
1549     10
1550     10
1551     10
1552     10
1553     10
1554     10
1555     10
1556     10
       ... 
782      10
781      10
775      10
799      10
780      10
779      10
776      10
777      10
778      10
800      10
784      11
1068     11
1662     11
342      15
1663     16
1165     20
1598     20
1433     40
1202     50
1351     50
1274     50
433      70
1843     80
1254     80
1228     90
1635    110
1779    120
1634    130
902     150
1120    170
Name: rating_denominator, Length: 2356, dtype: int64

In [29]:
len(tweet[(tweet['doggo'] == 'doggo')]) + len(tweet[(tweet['floofer'] == 'floofer')])  + len(tweet[(tweet['pupper'] == 'pupper')]) 
+ len(tweet[(tweet['puppo'] == 'puppo')]) 

30

#### Quality
##### `tweet` table
- in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, and retweeted_status_user_id are int not float
- Erroneous datatypes (retweeted_status_timestamp, timestamp, text, and source columns) 
- source has four categories
- name sometimes lower case, captial letter other times
- Missing expanded_urls (2297 instead of 2356)
- Missing in_reply_to_status_id and in_reply_to_user_id (78 instead of 2356)
- Missing retweeted_status_id, retweeted_status_id, retweeted_status_id (181 instead of 2356)
- actually, by, his, my, one, the, very are not dog name
- Multiple records for 'a, an' as a name
- Some rating_numerator and rating_denominator are extremely high
- Some rows have rating_numerator and rating_denominator but do not have expanded_urls

##### `image` table
- Duplicates in jpg_url
- Lowercase p1, p2, p3
- Missing records (2075 instead of 2356)
- Erroneous datatypes (jpg_url column)
- Nulls represented as dashes (-) in auralin and novodra columns

##### `info` table
- Missing records (2069 instead of 2075)

#### Tidiness
- Merge doggo, pupper, puppo, and floofer columns into one column in `tweet` table
- `info` table should be part of the `image` table

## Clean

In [ ]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

### Missing Data

<font color='red'>Complete the following two "Missing Data" **Define, Code, and Test** sequences after watching the *"Address Missing Data First"* video.</font>

#### `treatments`: Missing records (280 instead of 350)

##### Define
*Your definition here. Note: the missing `treatments` records are stored in a file named `treatments_cut.csv`, which you can see in this Jupyter Notebook's dashboard (click the **jupyter** logo in the top lefthand corner of this Notebook). Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html) for the function used in the solution.*

Import the cut treatments into a DataFrame and concatenate it with the original treatments DataFrame.

##### Code

In [ ]:
# Your cleaning code here
treatments_cut = pd.read_csv('treatments_cut.csv')
treatments_clean = pd.concat([treatments_clean, treatments_cut],
                             ignore_index=True)

##### Test

In [ ]:
# Your testing code here
treatments_clean.head()

#### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)
*Note: the "Inaccurate HbA1c changes (leading 4s mistaken as 9s)" observation, which is an accuracy issue and not a completeness issue, is included in this header because it is also fixed by the cleaning operation that fixes the missing "Missing HbA1c changes" observation. Multiple observations in one **Define, Code, and Test** header occurs multiple times in this notebook.*

##### Define
*Your definition here.*

Recalculate the hba1c_change column: hba1c_start minus hba1c_end.

##### Code

In [ ]:
# Your cleaning code here
treatments_clean.hba1c_change = (treatments_clean.hba1c_start - 
                                 treatments_clean.hba1c_end)

##### Test

In [ ]:
# Your testing code here
treatments_clean.hba1c_change.head()

### Tidiness

<font color='red'>Complete the following four "Tidiness" **Define, Code, and Test** sequences after watching the *"Cleaning for Tidiness"* video.</font>

#### Contact column in `patients` table contains two variables: phone number and email

##### Define
*Your definition here. Hint 1: use regular expressions with pandas' [`str.extract` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.extract.html). Here is an amazing [regex tutorial](https://regexone.com/). Hint 2: [various phone number regex patterns](https://stackoverflow.com/questions/16699007/regular-expression-to-match-standard-10-digit-phone-number). Hint 3: [email address regex pattern](http://emailregex.com/), which you might need to modify to distinguish the email from the phone number.*

Extract the phone number and email variables from the contact column using regular expressions and pandas' str.extract method. Drop the contact column when done.

##### Code

In [ ]:
# Your cleaning code here
patients_clean['phone_number'] = patients_clean.contact.str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})', expand=True)

# [a-zA-Z] to signify emails in this dataset all start and end with letters
patients_clean['email'] = patients_clean.contact.str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])', expand=True)

# Note: axis=1 denotes that we are referring to a column, not a row
patients_clean = patients_clean.drop('contact', axis=1)

##### Test

In [ ]:
# Your testing code here

# Confirm contact column is gone
list(patients_clean)

In [ ]:
patients_clean.phone_number.sample(25)

In [ ]:
# Confirm that no emails start with an integer (regex didn't match for this)
patients_clean.email.sort_values().head()

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

##### Define
*Your definition here. Hint: use pandas' [melt function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html) and [`str.split()` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.split.html). Here is an excellent [`melt` tutorial](https://deparkes.co.uk/2016/10/28/reshape-pandas-data-with-melt/).*

Melt the auralin and novodra columns to a treatment and a dose column (dose will still contain both start and end dose at this point). Then split the dose column on ' - ' to obtain start_dose and end_dose columns. Drop the intermediate dose column.

##### Code

In [ ]:
# Your cleaning code here
treatments_clean = pd.melt(treatments_clean, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')
treatments_clean = treatments_clean[treatments_clean.dose != "-"]
treatments_clean['dose_start'], treatments_clean['dose_end'] = treatments_clean['dose'].str.split(' - ', 1).str
treatments_clean = treatments_clean.drop('dose', axis=1)

##### Test

In [ ]:
# Your testing code here
treatments_clean.head()

#### Adverse reaction should be part of the `treatments` table

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for the function used in the solution.*

Merge the adverse_reaction column to the treatments table, joining on given name and surname.

##### Code

In [ ]:
# Your cleaning code here
treatments_clean = pd.merge(treatments_clean, adverse_reactions_clean,
                            on=['given_name', 'surname'], how='left')

##### Test

In [ ]:
# Your testing code here
treatments_clean

#### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for one function used in the solution and [tutorial](http://erikrood.com/Python_References/dropping_rows_cols_pandas.html) for another function used in the solution.*

Adverse reactions table is no longer needed so ignore that part. Isolate the patient ID and names in the patients table, then convert these names to lower case to join with treatments. Then drop the given name and surname columns in the treatments table (so these being lowercase isn't an issue anymore).

##### Code

In [ ]:
# Your cleaning code here
id_names = patients_clean[['patient_id', 'given_name', 'surname']]
id_names.given_name = id_names.given_name.str.lower()
id_names.surname = id_names.surname.str.lower()
treatments_clean = pd.merge(treatments_clean, id_names, on=['given_name', 'surname'])
treatments_clean = treatments_clean.drop(['given_name', 'surname'], axis=1)

##### Test

In [ ]:
# Your testing code here
treatments_clean

In [ ]:
# Patient ID should be the only duplicate column
all_columns = pd.Series(list(patients_clean) + list(treatments_clean))
all_columns[all_columns.duplicated()]

### Quality

<font color='red'>Complete the remaining "Quality" **Define, Code, and Test** sequences after watching the *"Cleaning for Quality"* video.</font>

#### Zip code is a float not a string and Zip code has four digits sometimes

##### Define
*Your definition here. Hint: see the "Data Cleaning Process" page.*

Convert the zip code column's data type from a float to a string using astype, remove the '.0' using string slicing, and pad four digit zip codes with a leading 0.

##### Code

In [ ]:
# Your cleaning code here
patients_clean.zip_code = patients_clean.zip_code.astype(str).str[:-2].str.pad(5, fillchar = '0')
# Reconvert NaNs entries that were converted to '0000n' by code above
patients_clean.zip_code = patients_clean.zip_code.replace('0000n', np.nan)

##### Test

In [ ]:
# Your testing code here
patients_clean.zip_code.head()

#### Tim Neudorf height is 27 in instead of 72 in

##### Define
*Your definition here.*
Replace height for rows in the patients table that have a height of 27 in (there is only one) with 72 in.

##### Code

In [ ]:
# Your cleaning code here
patients_clean.height = patients_clean.height.replace(27, 72)

##### Test

In [ ]:
# Your testing code here
# Should be empty
patients_clean[patients_clean.height == 27]

#### Full state names sometimes, abbreviations other times

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_apply_operations_to_dataframes.html) for method used in solution.*

Apply a function that converts full state name to state abbreviation for California, New York, Illinois, Florida, and Nebraska.

##### Code

In [ ]:
# Your cleaning code here
# Mapping from full state name to abbreviation
state_abbrev = {'California': 'CA',
                'New York': 'NY',
                'Illinois': 'IL',
                'Florida': 'FL',
                'Nebraska': 'NE'}

# Function to apply
def abbreviate_state(patient):
    if patient['state'] in state_abbrev.keys():
        abbrev = state_abbrev[patient['state']]
        return abbrev
    else:
        return patient['state']
    
patients_clean['state'] = patients_clean.apply(abbreviate_state, axis=1)


##### Test

In [ ]:
# Your testing code here
patients_clean.state.value_counts()

#### Dsvid Gustafsson

##### Define
*Your definition here.*

Replace given name for rows in the patients table that have a given name of 'Dsvid' with 'David'.

##### Code

In [ ]:
# Your cleaning code here
patients_clean.given_name = patients_clean.given_name.replace('Dsvid', 'David')

##### Test

In [ ]:
# Your testing code here
patients_clean[patients_clean.surname == 'Gustafsson']

#### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

##### Define
*Your definition here. Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.astype.html) for one method used in solution, [documentation page](http://pandas.pydata.org/pandas-docs/version/0.20/generated/pandas.to_datetime.html) for one function used in the solution, and [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.strip.html) for another method used in the solution.*

Convert assigned sex and state to categorical data types. Zip code data type was already addressed above. Convert birthdate to datetime data type. Strip the letter 'u' in start dose and end dose and convert those columns to data type integer.

##### Code

In [ ]:
# Your cleaning code here
# To category
patients_clean.assigned_sex = patients_clean.assigned_sex.astype('category')
patients_clean.state = patients_clean.state.astype('category')

# To datetime
patients_clean.birthdate = pd.to_datetime(patients_clean.birthdate)

# Strip u and to integer
treatments_clean.dose_start = treatments_clean.dose_start.str.strip('u').astype(int)
treatments_clean.dose_end = treatments_clean.dose_end.str.strip('u').astype(int)

##### Test

In [ ]:
# Your testing code here
patients_clean.info()

#### Multiple phone number formats

##### Define
*Your definition here. Hint: helpful [Stack Overflow answer](https://stackoverflow.com/a/123681).*

Strip all " ", "-", "(", ")", and "+" and store each number without any formatting. Pad the phone number with a 1 if the length of the number is 10 digits (we want country code).

##### Code

In [ ]:
# Your cleaning code here
patients_clean.phone_number = patients_clean.phone_number.str.replace(r'\D+', '').str.pad(11, fillchar='1')

##### Test

In [ ]:
# Your testing code here
patients_clean.phone_number.head()

#### Default John Doe data

##### Define
*Your definition here. Recall that it is assumed that the data that this John Doe data displaced is not recoverable.*

Remove the non-recoverable John Doe records from the patients table.

##### Code

In [ ]:
# Your cleaning code here
patients_clean = patients_clean[patients_clean.surname != 'Doe']

##### Test

In [ ]:
# Your testing code here
# Should be no Doe records
patients_clean.surname.value_counts()

In [ ]:
# Should be no 123 Main Street records
patients_clean.address.value_counts()

#### Multiple records for Jakobsen, Gersten, Taylor

##### Define
*Your definition here.*

Remove the Jake Jakobsen, Pat Gersten, and Sandy Taylor rows from the patients table. These are the nicknames, which happen to also not be in the treatments table (removing the wrong name would create a consistency issue between the patients and treatments table). These are all the second occurrence of the duplicate. These are also the only occurences of non-null duplicate addresses.

##### Code

In [ ]:
# Your cleaning code here
# tilde means not: http://pandas.pydata.org/pandas-docs/stable/indexing.html#boolean-indexing
patients_clean = patients_clean[~((patients_clean.address.duplicated()) & patients_clean.address.notnull())]

##### Test

In [ ]:
# Your testing code here
patients_clean[patients_clean.surname == 'Jakobsen']

In [ ]:
patients_clean[patients_clean.surname == 'Gersten']

In [ ]:
patients_clean[patients_clean.surname == 'Taylor']

#### kgs instead of lbs for Zaitseva weight

##### Define
*Your definition here.*

Use advanced indexing to isolate the row where the surname is Zaitseva and convert the entry in its weight field from kg to lbs.

##### Code

In [ ]:
# Your cleaning code here
weight_kg = patients_clean.weight.sort_values()[0]
mask = patients_clean.surname == 'Zaitseva'
column_name = 'weight'
patients_clean.loc[mask, column_name] = weight_kg * 2.20462

##### Test

In [ ]:
# Your testing code here
# 48.8 shouldn't be the lowest anymore
patients_clean.weight.sort_values()